### ISSUE TITLE
How can I crop segmented object from image 

### ISSUE THREAD

**Issue by Akansel-AKYOL** [ultralytics/yolov5#10122]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

Hi,
I want to get crop segmented objects from image by using yolov5 segmentation model. --save-crop argument in predict.py crops objects in bounding box. But I want only segmented part without background. Is there any way?

Predicted image:

![image](https://user-images.githubusercontent.com/72917780/201114179-b5f16045-6183-41c1-873e-6757b8c3a5a7.png)

--save-crop argument crops it as a bounding box as below:
 
![bus2](https://user-images.githubusercontent.com/72917780/201114862-b0fef73c-af3d-4049-a7a6-d5e2aa3e7448.jpeg)

The shape I want is:
![image](https://user-images.githubusercontent.com/72917780/201118714-cec80430-128c-4529-a8f0-b77adef8ca1f.png)

**Comment by lakshaydulani** [comment]
update the masks function in **plots.py**

`def masks(self, masks, colors, im_gpu=None, alpha=0.5):
        alpha=0
        """Plot masks at once.
        Args:
            masks (tensor): predicted masks on cuda, shape: [n, h, w]
            colors (List[List[Int]]): colors for predicted masks, [[r, g, b] * n]
            im_gpu (tensor): img is in cuda, shape: [3, h, w], range: [0, 1]
            alpha (float): mask transparency: 0.0 fully transparent, 1.0 opaque
        """
        if self.pil:
            # convert to numpy first
            self.im = np.asarray(self.im).copy()
        if im_gpu is None:
            
            # Add multiple masks of shape(h,w,n) with colors list([r,g,b], [r,g,b], ...)
            if len(masks) == 0:
                return
            if isinstance(masks, torch.Tensor):
                masks = torch.as_tensor(masks, dtype=torch.uint8)
                masks = masks.permute(1, 2, 0).contiguous()
                masks = masks.cpu().numpy()
            # masks = np.ascontiguousarray(masks.transpose(1, 2, 0))
            masks = scale_image(masks.shape[:2], masks, self.im.shape)
            masks = np.asarray(masks, dtype=np.float32)
            colors = np.asarray(colors, dtype=np.float32)  # shape(n,3)
            s = masks.sum(2, keepdims=True).clip(0, 1)  # add all masks together
            masks = (masks @ colors).clip(0, 255)  # (h,w,n) @ (n,3) = (h,w,3)
            self.im[:] = masks * alpha + self.im * (1 - s * alpha)
        else:
            
            if len(masks) == 0:
                self.im[:] = im_gpu.permute(1, 2, 0).contiguous().cpu().numpy() * 255
            colors = torch.tensor(colors, device=im_gpu.device, dtype=torch.float32) / 255.0
            colors = colors[:, None, None]  # shape(n,1,1,3)
                        
            masks = masks.unsqueeze(3)  # shape(n,h,w,1)
            masks_color = masks * (colors * alpha)  # shape(n,h,w,3)

            inv_alph_masks = (1 - masks * alpha).cumprod(0)  # shape(n,h,w,1)
            mcs = (masks_color * inv_alph_masks).sum(0) * 2  # mask color summand shape(n,h,w,3)

            im_gpu = im_gpu.flip(dims=[0])  # flip channel
            im_gpu = im_gpu.permute(1, 2, 0).contiguous()  # shape(h,w,3)
            # im_gpu = im_gpu * inv_alph_masks[-1] + mcs
            im_gpu = im_gpu * masks[-1]
            im_mask = (im_gpu * 255).byte().cpu().numpy()
            self.im[:] = scale_image(im_gpu.shape, im_mask, self.im.shape)
        if self.pil:
            # convert im back to PIL and update draw
            self.fromarray(self.im)`

**Comment by Akansel-AKYOL** [comment]
> update the masks function in **plots.py**
> 
> def masks(self, masks, colors, im_gpu=None, alpha=0.5): alpha=0 """Plot masks at once. Args: masks (tensor): predicted masks on cuda, shape: [n, h, w] colors (List[List[Int]]): colors for predicted masks, [[r, g, b] * n] im_gpu (tensor): img is in cuda, shape: [3, h, w], range: [0, 1] alpha (float): mask transparency: 0.0 fully transparent, 1.0 opaque """ if self.pil: # convert to numpy first self.im = np.asarray(self.im).copy() if im_gpu is None: # Add multiple masks of shape(h,w,n) with colors list([r,g,b], [r,g,b], ...) if len(masks) == 0: return if isinstance(masks, torch.Tensor): masks = torch.as_tensor(masks, dtype=torch.uint8) masks = masks.permute(1, 2, 0).contiguous() masks = masks.cpu().numpy() # masks = np.ascontiguousarray(masks.transpose(1, 2, 0)) masks = scale_image(masks.shape[:2], masks, self.im.shape) masks = np.asarray(masks, dtype=np.float32) colors = np.asarray(colors, dtype=np.float32) # shape(n,3) s = masks.sum(2, keepdims=True).clip(0, 1) # add all masks together masks = (masks @ colors).clip(0, 255) # (h,w,n) @ (n,3) = (h,w,3) self.im[:] = masks * alpha + self.im * (1 - s * alpha) else:
> 
> ```
>         if len(masks) == 0:
>             self.im[:] = im_gpu.permute(1, 2, 0).contiguous().cpu().numpy() * 255
>         colors = torch.tensor(colors, device=im_gpu.device, dtype=torch.float32) / 255.0
>         colors = colors[:, None, None]  # shape(n,1,1,3)
>                     
>         masks = masks.unsqueeze(3)  # shape(n,h,w,1)
>         masks_color = masks * (colors * alpha)  # shape(n,h,w,3)
> 
>         inv_alph_masks = (1 - masks * alpha).cumprod(0)  # shape(n,h,w,1)
>         mcs = (masks_color * inv_alph_masks).sum(0) * 2  # mask color summand shape(n,h,w,3)
> 
>         im_gpu = im_gpu.flip(dims=[0])  # flip channel
>         im_gpu = im_gpu.permute(1, 2, 0).contiguous()  # shape(h,w,3)
>         # im_gpu = im_gpu * inv_alph_masks[-1] + mcs
>         im_gpu = im_gpu * masks[-1]
>         im_mask = (im_gpu * 255).byte().cpu().numpy()
>         self.im[:] = scale_image(im_gpu.shape, im_mask, self.im.shape)
>     if self.pil:
>         # convert im back to PIL and update draw
>         self.fromarray(self.im)
> ```

Thank you @lakshaydulani but it didn't work as i wanted

**Comment by factorLee** [comment]
did you solve it?

**Comment by glenn-jocher** [comment]
@factorLee hello,

I apologize but I'm not sure which specific issue you're referring to. If you could clarify or provide more details, I'd be happy to try and assist you in finding a solution. 

Thank you for using YOLOv5, and please don't hesitate to ask if you have any further questions or concerns.

Best regards,

**Comment by factorLee** [comment]
@glenn-jocher Hi!,
What I want is like the questioner.
I want to segment an object in an image using a custom model obtained through the segment/train.py file and crop the segmented part. Instead of cropping to four sides of the bbox, you want to precisely crop the segmented object area.

In the process of loading the model and making predictions, the variable called masks contains the binary value of the predicted object area, but the h and w values ​​are different from the original image, so it seems personally difficult to calculate the crop.

how can i solve it?

**Comment by glenn-jocher** [comment]
Hello @factorLee,

Thank you for your question! 

To crop the segmented object area, one possible approach is to use the masks variable obtained during the inference process to create a bounding box that precisely fits the segmented object. You can use the binary mask to obtain the coordinates of the minimum enclosing rectangle that bounds the object. Once you have the coordinates of the bounding box, you can easily crop the image to that area.

To get the coordinates of the minimum enclosing rectangle from the binary mask, you can use the **cv2** module in Python. Here's an example code for obtaining the bounding box coordinates:

```python
import cv2

# Load binary mask and original image
binary_mask = masks[0].cpu().numpy() # convert tensor to numpy array
original_img = cv2.imread("path/to/your/image.jpg") # read original image from file

# Get coordinates of minimum enclosing rectangle
(x, y, w, h) = cv2.boundingRect(binary_mask.astype('uint8'))

# Crop image to object area
object_area = original_img[y:y+h, x:x+w]
```

This code should give you a cropped image that precisely fits the segmented object area defined by the binary mask. Note that in this example, we assume that the binary mask is a numpy array obtained from the tensor returned by YOLOv5.

I hope this helps! Let me know if you have any questions or need further assistance.

Best regards,

**Comment by factorLee** [comment]
@glenn-jocher 
Thank you for your kind reply!
 However, what I want is to multiply the object area information (boolean type) of masks obtained through model prediction by the original image (broadcast), so that the image consists of a black background other than the object.
 That is, the area of ​​the predicted object must retain the pixel values ​​of the original image and otherwise have a black color (rgb=0).

 I am curious about cropping according to the shape of the object rather than the square box for the minimum area you mentioned.
 

For example, the result should look like the image below:
![image](https://github.com/ultralytics/yolov5/assets/53315807/e7708884-fb33-4b15-a9cd-e346af88b0f3)

Best regards,

**Comment by glenn-jocher** [comment]
Hello @factorLee,

Thank you for clarifying your requirements. To apply a Boolean mask obtained from model prediction to the original image and crop according to the shape of the object, you can do the following:

```python
# Load binary mask and original image
binary_mask = masks[0].cpu().numpy() # convert tensor to numpy array
original_img = Image.open("path/to/your/image.jpg") # read original image from file

# Create black background for mask applying to original image
background = Image.new('RGB', original_img.size, color=(0, 0, 0))

# Apply mask to original image to retain original pixel values in object area
object_area = Image.composite(original_img, background, Image.fromarray(binary_mask.astype('uint8')))

# Get coordinates of object bounding box
contours, _ = cv2.findContours(binary_mask.astype('uint8'), cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)
(x, y, w, h) = cv2.boundingRect(contours[0])

# Crop image to object area shape
cropped_object = object_area.crop((x, y, x+w, y+h))
```

This code creates a new image with a black background, where the same object area in the original image is retained based on the binary mask. Then, it obtains the coordinates of the object area in the image using a contour detection method. Finally, it crops the object area to the exact shape of the object using these coordinates.

I hope this solution works for you! Please let me know if you have any further questions or need further assistance.

Best regards,
